File: src/dorcs/meta_data.rs
use regex::Regex;

#[derive(Debug, Clone)]
pub struct MetaData {
    pub title: String,
    pub position: Option<i32>,
}

impl MetaData {
    pub fn from_string(content: &String) -> MetaData {
        Self::parse_metadata(content)
    }

    // Parses the metadata from a markdown file.
    fn parse_metadata(raw_content: &String) -> MetaData {
        let re = Regex::new(r"(?s)^---\s*(.*?)\s*---").unwrap();
        let mut title = String::new();
        let mut position: Option<i32> = None;

        if let Some(cap) = re.captures_iter(raw_content.as_str()).next() {
            let meta_str = &cap[1];

            for line in meta_str.lines() {
                let parts: Vec<&str> = line.splitn(2, ':').collect();
                if parts.len() == 2 {
                    let key = parts[0].trim();
                    let value = parts[1].trim();

                    // Match the key and update the struct fields accordingly
                    match key {
                        "title" => title = value.to_string(),
                        "position" => position = value.parse::<i32>().ok(),
                        _ => {} // Ignore unknown keys
                    }
                }
            }
        }

        MetaData { title, position }
    }

    pub fn remove_meta_data(content: &String) -> String {
        let re = Regex::new(r"(?s)^---\s*(.*?)\s*---").unwrap();
        let content = re.replace(content, "");
        content.to_string()
    }
}

File: pkg/dorcs/meta_data.go
package dorcs 

import (
	"regexp"
	"strconv"
	"strings"
)

type MetaData struct {
	Title    string
	Position int
}

// Parses the metadata from a markdown file.
func ParseMetadata(content string) *MetaData {
	re := regexp.MustCompile(`(?s)^---\s*(.*?)\s*---`)
	metaData := &MetaData{}

	if cap := re.FindStringSubmatch(content); len(cap) > 0 {
		metaStr := cap[1]
		lines := strings.Split(metaStr, "\n")

		for _, line := range lines {
			parts := strings.SplitN(line, ":", 2)
			if len(parts) == 2 {
				key := strings.TrimSpace(parts[0])
				value := strings.TrimSpace(parts[1])

				// Match the key and update the struct fields accordingly
				switch key {
				case "title":
					metaData.Title = value
				case "position":
					if pos, err := strconv.ParseInt(value, 10, 64); err == nil {
						metaData.Position = int(pos)
					}
				default:
					// Ignore unknown keys
				}
			}
		}
	}

	return metaData
}

// Remove metadata from the content
func RemoveMetadata(content string) string {
	re := regexp.MustCompile(`(?s)^---\s*(.*?)\s*---`)
	return re.ReplaceAllString(content, "")
}
